### PR TITLE
Feat/filter

### PR DESCRIPTION
**Pull request by ahudson7**
Implementations in this branch:
- state variable filter
- waveform visualizer
- UI elements for all
- restructured UI to accommodate new elements

### CODE REVIEW

File: Source/PluginEditor.cpp [hunk 1]
```diff
@@ -15,25 +15,57 @@ SubsynthAudioProcessorEditor::SubsynthAudioProcessorEditor (SubsynthAudioProcess
...
-        
+            
+    filterSelect.addItem("Low Pass", 1);
+    filterSelect.addItem("Band Pass", 2);
+    filterSelect.addItem("High Pass", 3);
+    filterSelect.onChange = [this] { filterChanged(); };
+    filterSelect.setSelectedId(1);
+
+    filterCutoff.setSliderStyle(juce::Slider::SliderStyle::LinearHorizontal);
+    filterCutoff.setRange(20.0f, 20000.0f, 0.01f);
```
Comment (anwils2) [inline Source/PluginEditor.cpp]: ~I think its worth looking into making this slider value scale non-linearly. Something like the first half of the bar being 1-5k Hz and other half the remaining 15k Hz. I don't think its worth holding up this PR and can always be added later.~

Sometimes JUCE is ez:
```suggestion
    filterCutoff.setRange(20.0f, 20000.0f, 0.01f);
    filterCutoff.setSkewFactorFromMidPoint(5000);
```

File: Source/CustomVoice.cpp [hunk 2]
```diff
@@ -43,9 +43,16 @@ void CustomVoice::prepareToPlay(double sampleRate, int samplesPerBlock, int numO
     spec.sampleRate = sampleRate;
     spec.numChannels = numOutputChannels;
 
+    sampleRateHolder = sampleRate;
+
     juce::ADSR::Parameters initADSR{
         0.1f, 0.1f, 0.1f, 0.1f
     };
+    spec.numChannels = numOutputChannels;
```
Comment (anwils2) [inline Source/CustomVoice.cpp]: Is this a duplicate of line 44?
```suggestion

```

File: Source/PluginEditor.cpp [hunk 3]
```diff
@@ -15,25 +15,57 @@ SubsynthAudioProcessorEditor::SubsynthAudioProcessorEditor (SubsynthAudioProcess
...
+    filterCutoff.setValue(10000.0f);
+    filterCutoff.setTextBoxStyle(juce::Slider::NoTextBox, false, 0, 0);
+    filterCutoff.setPopupDisplayEnabled(true, true, this);
+    filterCutoff.onDragEnd = [this] { filterChanged(); };
```
Comment (anwils2) [inline Source/PluginEditor.cpp:43]: Minor nit cosmetic stuff
```suggestion
    filterCutoff.setTextValueSuffix(" Hz");
    
```

**Review by anwils2 — APPROVED** [review]

File: Source/PluginEditor.cpp [hunk 4]
```diff
@@ -15,25 +15,59 @@ SubsynthAudioProcessorEditor::SubsynthAudioProcessorEditor (SubsynthAudioProcess
...
-        
+            
+    filterSelect.addItem("Low Pass", 1);
+    filterSelect.addItem("Band Pass", 2);
+    filterSelect.addItem("High Pass", 3);
+    filterSelect.onChange = [this] { filterChanged(); };
+    filterSelect.setSelectedId(1);
+
+    filterCutoff.setSliderStyle(juce::Slider::SliderStyle::LinearHorizontal);
+    filterCutoff.setRange(20.0f, 20000.0f, 0.01f);
```
Comment (anwils2) [inline Source/PluginEditor.cpp]: Sorry 1 other very tiny change. I don't think we need this much resolution. Filter setting down to 1 Hz is probably better than showing the decimal.
```suggestion
    filterCutoff.setRange(20.0f, 20000.0f, 1.0f);
```

File: Source/PluginEditor.cpp [hunk 5]
```diff
@@ -15,25 +15,59 @@ SubsynthAudioProcessorEditor::SubsynthAudioProcessorEditor (SubsynthAudioProcess
...
     
     // Add listeners
     waveSelect.addListener(this);
+    filterSelect.addListener(this);
+    filterCutoff.addMouseListener(this, true);
+    filterRes.addMouseListener(this, true);
+
+    // Waveform Visualiser
+    addAndMakeVisible(&p.wfVisualiser);
+    p.wfVisualiser.setBounds(10, 360, getWidth() - 20, 200); // add to resized() below - figure out how to access p there?
```
Comment (robinjsu) [inline Source/PluginEditor.cpp]: `p` equivalent to the `audioProcessor`, so you can access by invoking `audioProcessor.wfVisualiser.setBounds(...)` in `resized()`

**Pull request merged**